### PR TITLE
nixos/nginx: allow enabling QUIC packet routing using eBPF

### PR DESCRIPTION
**Pull request by Izorkin**
## Description of changes
Allow enabling QUIC packet routing using eBPF.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -1224,8 +1246,8 @@ in
         # New file permissions
         UMask = "0027"; # 0640 / 0750
         # Capabilities
-        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
-        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ] ++ optionals cfg.enableQuicBPF [ "CAP_SYS_ADMIN" "CAP_NET_ADMIN" ];
+        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ] ++ optionals cfg.enableQuicBPF [ "CAP_SYS_ADMIN" "CAP_NET_ADMIN" ];
```
Comment (Mic92) [inline nixos/modules/services/web-servers/nginx/default.nix:1250]: Those are quite powerful capabilities. Does nginx drop them in its webserver?

Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:1250]: Without these features, eBPF-enabled nginx will not work :( . Nginx does not use CAP_BPF.

Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:1250]: @arut is it possible to implement use of eBPF only with Capability "CAP_BPF"?

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 2]
```diff
@@ -1125,6 +1139,14 @@ in
         '';
       }
 
+      {
+        assertion = cfg.package.pname != "nginxQuic" -> !(cfg.enableQuicBPF);
+        message = ''
+          services.nginx.enableQuicBPF requires using nginxQuic package,
+          which can be achieved by setting `services.nginx.package = pkgs.nginxQuic;`.
+        '';
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:1147]: Should we also assert that at least nginx 1.25 is used? 1.24 does not support this option despite having initial quic support

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/modules/services/web-servers/nginx/default.nix:1250]: > Does nginx drop them in its webserver?

I've run getcaps on a worker process and it does not.

```
 ➜ getpcaps 11070
11070: cap_net_bind_service,cap_net_admin,cap_sys_admin,cap_sys_resource,cap_bpf=eip
```

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 2 above)
Comment (Izorkin) [reply nixos/modules/services/web-servers/nginx/default.nix:1147]: This feature appeared long before version 1.24 in a separate nginx-quic branch.

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 3]
```diff
@@ -783,6 +787,16 @@ in
         '';
       };
 
+      enableQuicBPF = mkOption {
+        default = false;
+        type = types.bool;
+        description = lib.mdDoc ''
+          Enables routing of QUIC packets using eBPF. When enabled, this allows
+          to support QUIC connection migration. The directive is only supported
+          on Linux 5.7+.
```
Comment (Mic92) [inline nixos/modules/services/web-servers/nginx/default.nix:796]:
```suggestion
          on Linux 5.7+. Note that enabling this option will make nginx run with extended capabilities that are usually limited to processes running as root namely `CAP_SYS_ADMIN` and `CAP_NET_ADMIN`
```

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 1 above)
Comment (Mic92) [reply nixos/modules/services/web-servers/nginx/default.nix:1250]: If this is how people like to run their webserver these days, so it be. Let's at least put a warning in.... next up we are going to modprobe a nginx.ko into the kernel for more performance.

**Review by RaitoBezarius — APPROVED** [review]

**Pull request merged**